htdocs/compta/facture/class/facture.class.php

```diff
@@ -2808,6 +2809,26 @@ public function validate($user, $force_number = '', $idwarehouse = 0, $notrigger
 				}
 			}
 
+			// Generate PDF
```

Comment:
I am not sure it is a good idea to merge 2 operations that are so different;
- validate is a low level database change
- generation of PDF is a heavy IO process (not necessarly done when validate is inside a batch that make other transformations after the validate for example).

So we must keep   validate()   and   generateDocument()   as   2 different operations not necessarly linked.

But we can add an option into the validate method to add the action so we will have both actions in same method if we provide parameter to 1, and we can still ignore document generation if we need. Why not just adding a test on

if (!empty($pdf)) {

This is our option.
Also please rename $pdf into $arraytobuilddoc   (doc is not always a PDF)

Reply:
Then generation of doc is an operation not related to database, so it must be done completely at end, so after the trigger  BILL_VALIDATE that is still inside the database context.

Reply:
the mistake with this that when validating no pdf or doc are joined with templated email because it was generated after.

all modern ERP/CRM can joined pdf or doc immediately  with notification in templated emails

Reply:
@eldy have you some idea for notifications with doc (pdf or odt) 
the mistake is for validation because doc generation is after....

Reply:
You can join pdf and doc immediatly. This is inside generateDocument(). For example, the module concatpdf can join an external pdf each time a pdf is modified.
For notification, we can imagine that the notification check the doc is up to date and update it if not but only when it needs to attach the file to it (so the notification can call the generateDocument() for example in case it need it).

Reply:
good idea. I push a new PR